### ISSUE TITLE
Resize up not working when window bottom near screen bottom

### ISSUE THREAD

**Issue by etienned** [Hammerspoon/hammerspoon#217]
I'm new to Hammerspoon and when doing the tutorial I found that when the bottom of a window is near the bottom of the screen I'm no longer able to resize the window vertically. For example if I hit `cmd-alt-ctrl-Left`, so the window is size to half screen on the left and then I press`cmd-alt-ctrl-Up` to trigger this function:

```lua
hs.hotkey.bind({"cmd", "alt", "ctrl"}, "Up", function()
  local win = hs.window.focusedWindow()
  local f = win:frame()
  local s = win:size()

  s.h = s.h - 10
      
  win:setSize(s)
end)
```

nothing happen. This function works correctly if the bottom of the window is at least at 18px of the screen border. I have the same problem if I use `win:setFrame(f)` instead of `win:setSize(s)`.

I also saw that if I change this function to resize 1px at the time (`s.h = s.h - 1`) the function stop working at around 26px from the screen bottom (instead of 18px). Also with resize set to 1px I'm not able to resize terminal windows.

I'm on Yosemite 10.10.1 with Hammerspoon 0.9.21.

**Comment by cmsj** [comment]
I can reproduce this. I noticed that if I try to resize the window by hand, I have to drag a little way before it actually resizes. It seems like there's a stickiness to windows when they're at the bottom of the screen.

You could check if the bottom edge of the window is within that sticky zone at of the screen, and then adjust the amount you resize, to get it out of the sticky zone (optionally then resizing it back down to the point you actually wanted, if you don't mind the window looking like it's jumping around).

I have a horrible feeling that this is going to be very hard to work around in the code, because I'm not aware of that sticky area being at all optional :(

**Comment by etienned** [comment]
You're right, there's stickiness to windows, but this stickiness is present on the menu side (top) and on the dock side (so it depends where's the dock). Honestly, I find this "feature" a bit a strange. I can't see the advantage of this stickiness that works only when sizing down/moving away from the side? It's not like a snap to edge.

Anyway, we'll have definitely to workaround this non-optional "feature". I found that the sticky zone is 24px large. So, if I use steps of 24px or greater, it works. When using the `win:setFrame(f)` method, the first move will jump, then after you will have the animation.

Concerning the Terminal.app case, I found that the minimum step size is in relation with the terminal font size. For example:
- font size 9pt -> 7px step
- font size 14pt -> 10px step
- font size 72pt -> 46px step

In Terminal.app, manual resizing works by step size of 1 line, so that explains the relation with font size. And that explains why there's no animation when resizing via Hammerspoon. So, no need to use `win:setFrame()` for Terminal.app.

**Comment by cmsj** [comment]
Oh, I totally forgot about the Terminal part. Sorry!

Yeah, they tend to resize in steps rather than smoothly. It's because terminals look super weird if there's a border around the edge because you resized it a few pixels less that is needed for another row/column.

Did you go with a quick check of the application name to vary the resize amount if it's a terminal?

**Comment by cmsj** [comment]
Also I think these "quirks" of the system may well be worth collecting in the docs, so I'm going to re-open the issue.

**Comment by lowne** [comment]
I haven't tested it yet, so this is basically a reminder for myself, but the convoluted workaround in `:setFrameInScreenBounds()` in #545 might fix this - if so, we *could* put it into `:setFrame()` (and just leave the "in screen bounds" part for the former method) if there are no nasty tradeoffs, or at the very least suggest its use in the docs for these tricky cases - although that wouldn't immediately solve the issue with downstream uses like `hs.grid` (see #290).
I'll investigate further.

**Comment by lowne** [comment]
#563 was merged and is in 0.9.41, `:setFrameInScreenBounds()` (as opposed to `:setFrame()`) should work around the stickiness; since I think there's a good chance the stickiness "feature" might behave differently in different OS X versions, it'd be nice to have confirmation. 

As for downstream, `hs.grid` and `hs.window.layout` use the new method everywhere; so what's left (again, assuming the current workaround actually works) is:
- [ ] `hs.layout`
- [ ] `hs.mjomatic`
- [ ] `hs.tabs`
- [ ] possibly `hs.drawing`'s container windows, if they're also affected (haven't tested)

**Comment by lowne** [comment]
The new `hs.window.setFrameCorrectness` should take care of all these issues; unless there still are some outstanding, I suppose this issue can be closed.
